### ISSUE TITLE
PMP ACT

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#9]
### Technical Group

Privileged Spec IC

### ratification-pkg

Priv 1.11

### Technical Liaison

Greg Favor

### Task Category

Arch Tests

### Task Sub Category

- [ ] gcc
- [ ] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [ ] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

3Q2023

### Statement of Work (SOW)

SOW: [link](https://docs.google.com/document/d/18Cdt_YblWPkTS9TLK-05j9SYoNKnwYz0QZFjo9EiRfQ/edit)

**SOW Signoffs:** (delete those not needed)
  - Task group liaison sign-off date:
  - Development partner sign-off date:
  - ACT SIG sign-off date (if ACT work):

### Waiver

- [X] Freeze
- [X] Ratification

**Comment by jjscheel** [comment]
[ACT PR](https://github.com/riscv-non-isa/riscv-arch-test/pull/284)

@UmerShahidengr, what is the status here?  It is still marked as being reviewed in your spreadsheet, but the PR seems to have bene closed and re-opened.

**Comment by jjscheel** [comment]
@UmerShahidengr, would you kindly let me know what github id we have for Abdul as well?  THANKS!

**Comment by jjscheel** [comment]
Update from Allen: 
> 1. Got the next version of the trap handler working and tested (With help for 10xe folks) and merged.
> It is now possible to write tests of virtual memory,  transition between protections mode easily, and write tests that can partition > permissions inside test code and data.
> 2. continue to partition and Test Format Spec into  ACT Developer Spec and ACT User Spec, and continue to get new chair up to speed
> 3. no
> 4. I don't have any dates....

**Comment by jjscheel** [comment]
Oops, previous post was for wrong item.  Pls ignore.

**Comment by jjscheel** [comment]
@Abdulwadoodd, any update here?  Thanks.

**Comment by UmerShahidengr** [comment]
Update ⇾ June 12th, 2023:
No update so far. We will come back to this project in the next week

**Comment by UmerShahidengr** [comment]
Update ⇾ July 11th, 2023
Not much work can be done here unless misaligned test cases are added. Coverpoints for PMP are getting developed but they are being tracked on different issue i.e, #13

**Comment by UmerShahidengr** [comment]
Update ⇾ August 15th, 2023
PMP coverpoints are ready and are available [here](https://github.com/riscv-software-src/riscv-ctg/pull/62) for review.
Furthermore, misaligned access tests are still missing, which will be added in this quarter.

**Comment by UmerShahidengr** [comment]
Update ⇾ September 12th, 2023
No update on this.

**Comment by UmerShahidengr** [comment]
Update ⇾ September 26th, 2023
No update on this.

**Comment by UmerShahidengr** [comment]
Update ⇾ October 10th, 2023
No update on this.

**Comment by UmerShahidengr** [comment]
Update ⇾ October 24th, 2023
No update on this.

**Comment by UmerShahidengr** [comment]
Update ⇾ November 28th, 2023
PMP coverpoints have been added. [Here](https://github.com/riscv-software-src/riscv-ctg/pull/91) is the link to the PR.

**Comment by UmerShahidengr** [comment]
Update ⇾ December 12th, 2023
The review process is going on smoothly.

**Comment by UmerShahidengr** [comment]
Update ⇾ January 30th, 2024
The review process is going on.

**Comment by UmerShahidengr** [comment]
Update ⇾ March 5th, 2024
@MuhammadHammad001 has added privilege coverpoint support in RISC-V ISAC ([PR here](https://github.com/riscv-software-src/riscv-isac/pull/80)), @allenjbaum has been working with @MuhammadHammad001 to write a translator for priv architecture which will help in writing priv arch coverpoints with better readability.
The coverpoint definitions are getting converted according to the new template, it will be completed in this quarter hopefully.

**Comment by UmerShahidengr** [comment]
Update ⇾ April 2nd, 2024
@MuhammadHammad001 has been working on PMP coverpoints, the PR will be completed before the next meeting

**Comment by jjscheel** [comment]
@UmerShahidengr, how are the cover points coming along?

**Comment by jjscheel** [comment]
I see old PR closed.  Is this the new one: [riscv-non-isa/riscv-arch-test #462](https://github.com/riscv-non-isa/riscv-arch-test/pull/462)?

**Comment by UmerShahidengr** [comment]
Yes. Thats the new one, and it is now complete with all tests and coverpoint definitions

**Comment by UmerShahidengr** [comment]
Update June 11th, 2024:
Coverpoint definitions and new template of writing coverpoints for Priv Architecture are still in review process, @allenjbaum has been reviewing this one and @MuhammadHammad001 has been working on it full time

**Comment by UmerShahidengr** [comment]
Update June 25th, 2024:
This task is still in review. @allenjbaum and @MuhammadHammad001 have been working on it.

**Comment by UmerShahidengr** [comment]
Update July 23rd, 2024:
This task is in review. @allenjbaum is spending some time reviewing it. Hopefully it will be completed soon

**Comment by UmerShahidengr** [comment]
Update August 6th, 2024:
The review process is complete, @allenjbaum has reviewed the PR, now it will be merged soon

**Comment by jjscheel** [comment]
Great news, @UmerShahidengr!

**Comment by UmerShahidengr** [comment]
Update October 29th, 2024:
Still in merging phase. Will be merged in the next release meeting in SIG ARCH. @allenjbaum has been asked to merge this one

**Comment by jjscheel** [comment]
ACT and Coverpoints PR merged.

**Comment by jjscheel** [comment]
SOW marked as "Signoff" stage.  

@allenjbaum, @UmerShahidengr, can you folks kindly review the SOW and confirm that we've completed the work?

**Comment by jjscheel** [comment]
@allenjbaum, @UmerShahidengr, please review and approve completion of this SOW.

**Comment by allenjbaum** [comment]
There are several issues that have been filed against these tests:
It doesn't work for anything except 4 byte granularity (#577, #580)
It doesn't work on a model that is M only or M-U only. (#582)
It is lacking a fence.i between storing into the instruction stream and
branching to it, and has an extra JAL making it test twice (#575)

#580 and #575 are being addressed (sic) under development.
#582 is new, relatively simple to fix, except there is a WARL behavior
configuration requirement that will require Sail changes
In general, it shouldn't matter, in the sense that if a test tries to go
from M->S with no Smode, it will go to M (for M-only) or either M or U (for
M-U),
and if tries to go to Umode in an M-only, it will also just stay in M-mode.
The tests should still work, but not as intended - they'll be redundant.
That's OK
But, these tests won't currently compile because they always define
rvtest_strap_routine - and that should only be defined if S-mode is
supported,

so the RVTEST_CASE macro has to be modified (there is an issue filed for
that (#584).
IF the model doesn't describe the WARL behavior (does it go to M or U if an
illegal value is written to MPP in an M-U model) then there will be a
mismatch.
Fixing that will require both Sail updates and a command line option to
define that behavior, though.
That should be part of the configuration issue, and I don't think we should
solve this now.
-----------------------
We should also be creating tests for RV32/64E - there have been
announcements of such a part.
A separate SOW to re-generate all tests to work with RV32/64E
implementations.
That (may) require new cgf files, but also need changes in every test.h
macro that uses an XREG larger than X15.
If we are clever about it, we can make this so that all new tests do this
automatically ( which involves ensure that signature base registers and
temp registers never go beyond X25,

Bottom line: RVTEST_CASE macro instantiation in each test should be
replaced ( might be a simple global search/replace), other issues are
currently under development.

**Comment by jjscheel** [comment]
Thanks, @allenjbaum.

@UmerShahidengr, please let me know when the issues have been resolved.

**Comment by UmerShahidengr** [comment]
Thank you for summarizing all reported issues @allenjbaum, we have been working on the next release in which all these issues will be resolved. The next version will be released this week.
CC @MuhammadHammad001

**Comment by UmerShahidengr** [comment]
PMP Version-2 has been released in this PR. All outstanding issues have been addressed.
PR Link: https://github.com/riscv-non-isa/riscv-arch-test/pull/603

**Comment by jjscheel** [comment]
Complete.
